c/tskit/trees.c

```diff
@@ -2103,22 +2126,635 @@ tsk_treeseq_sample_count_stat(const tsk_treeseq_t *self, tsk_size_t num_sample_s
+    tsk_size_t *restrict num_alleles = tsk_malloc(num_sites * sizeof(*num_alleles));
+    tsk_bit_array_t *restrict sample_sets
+        = tsk_calloc(num_sample_chunks * state_dim, sizeof(*sample_sets));
+    tsk_bit_array_t *allele_samples
+        = tsk_calloc(max_alleles * num_sample_chunks, sizeof(*allele_samples));
+    tsk_bit_array_t all_samples_bits[num_sample_chunks];
+
+    if (site_offsets == NULL || num_alleles == NULL || sample_sets == NULL
+        || allele_samples == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    // A future improvement could get a union of all sample sets
+    // instead of all samples
+    get_all_samples_bits(all_samples_bits, num_samples, num_sample_chunks);
+
+    tsk_bit_array_t *allele_samples_row;
+    tsk_size_t num_alleles_cumsum = 0;
+    for (tsk_size_t t = 0; t < self->num_trees; t++) {
+        for (tsk_size_t s = 0; s < self->tree_sites_length[t]; s++) {
+            tsk_id_t site_id = self->tree_sites[t][s].id;
+            // Initialize the allele_samples with all samples in the ancestral allele
+            allele_samples_row
+                = GET_2D_ROW(allele_samples, num_sample_chunks, num_alleles_cumsum);
+            tsk_add_bit_arrays(allele_samples_row, all_samples_bits, num_sample_chunks);
+            // Store the allele offset for each site
+            site_offsets[site_id] = num_alleles_cumsum;
+            num_alleles_cumsum += self->site_mutations_length[site_id] + 1;
+        }
+    }
+
+    ret = get_all_mutation_samples(
+        self, num_sample_chunks, site_offsets, num_alleles, &allele_samples);
+    if (ret != 0) {
+        goto out;
+    }
+
+    // Number of pairs w/ replacement (sites)
+    *result_size = (num_sites * (1 + num_sites)) >> (tsk_size_t) 1;
+    *result = tsk_calloc(*result_size * result_dim, sizeof(*result));
+
+    tsk_bit_array_t *sample_bits
+        = tsk_calloc(num_sample_chunks * state_dim, sizeof(*sample_bits));
+
+    if (sample_bits == NULL || result == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    sample_weights_to_bit_array(
+        sample_weights, state_dim, num_samples, num_sample_chunks, &sample_bits);
+    bool polarised = false;
+
+    if (options & TSK_STAT_POLARISED) {
+        polarised = true;
+    }
+
+    // TODO: implement windows!
+    tsk_size_t inner = 0;
+    tsk_size_t result_offset = 0;
+    for (tsk_size_t site_a = 0; site_a < num_sites; site_a++) {
+        for (tsk_size_t site_b = inner; site_b < num_sites; site_b++) {
+            ret = compute_general_two_site_stat_result(site_a, site_offsets[site_a],
+                site_b, site_offsets[site_b], num_sample_chunks, num_alleles,
+                allele_samples, state_dim, sample_bits, result_dim, f, f_params, norm_f,
+                polarised, &((*result)[result_offset]));
+            if (ret != 0) {
+                goto out;
+            }
+            result_offset += state_dim;
```

Comment:
This should be `result_dim`, no? (as written, `state_dim` always equals `result_dim`, but doing so doesn't make things simpler, I don't think?)

Reply:
Yes, you're right. I will change this.